File: programmers/old/level3/N_Queen.java
import java.util.*;
class Solution {
	int answer = 0;
	public int solution(int n) {
		int[] arr = new int[n];
		dfs(arr,0);
		return answer;
	}
	boolean check(int[] arr, int cnt){
		for(int i = 0; i < cnt; ++i) {
			if(arr[i] == arr[cnt]) {
				return false;
			}
			if(Math.abs(arr[i]-arr[cnt]) == Math.abs(i-cnt)) {
				return false;
			}
		}
		return true;
	}
	void dfs(int[] arr, int cnt) {
		if(cnt == arr.length) {
			answer++;
			return;
		}
		for(int i = 0; i < arr.length; ++i) {
			arr[cnt] = i;
			if(check(arr,cnt)) {
				dfs(arr,cnt+1);
			}
		}
	}
}
public class N_Queen {
	public static void main(String[] args) {
		int n =12;
		Solution sol = new Solution();
		System.out.println(sol.solution(n));
	}
}

File: programmers/python/level3/N-Queen.py
def solution(n):
    arr = []
    s = set()
    return dfs(0, 0, arr, n, 0, s)

def dfs(count, r, arr, n, answer, s):
    if count == n:
        answer += 1
        return answer
    for c in range(n):
        if is_valid(r,c,arr, s):
            arr.append([r,c])
            s.add(c)
            answer = dfs(count+1, r+1, arr, n, answer, s)
            s.remove(c)
            arr.pop()
    return answer

def is_valid(cr,cc,arr,s):
    if cc in s:
        return False
    for r, c in arr:
        if abs(r-cr) == abs(c-cc):
            return False
    return True